tt_metal/hw/inc/accessor/detail/args_location.hpp

```diff
@@ -0,0 +1,169 @@
+ *
+ */
+template <
+    bool RankCRTA_ = false,
+    bool NumBanksCRTA_ = false,
+    bool TensorShapeCRTA_ = false,
+    bool ShardShapeCRTA_ = false,
+    bool BankCoordsCRTA_ = false>
+struct ArgsLocation {
+    // Fetch locations of the arguments
+    static constexpr bool RankCRTA = RankCRTA_;
+    static constexpr bool NumBanksCRTA = NumBanksCRTA_;
+    static constexpr bool TensorShapeCRTA = TensorShapeCRTA_;
+    static constexpr bool ShardShapeCRTA = ShardShapeCRTA_;
+    static constexpr bool BankCoordsCRTA = BankCoordsCRTA_;
+
+    static constexpr bool RankStatic = !RankCRTA;
+    static constexpr bool NumBanksStatic = !NumBanksCRTA;
+    static constexpr bool TensorShapeStatic = !TensorShapeCRTA;
+    static constexpr bool ShardShapeStatic = !ShardShapeCRTA;
+    static constexpr bool BankCoordsStatic = !BankCoordsCRTA;
+};
+
+/**
+ * @brief Holds offsets for compile-time and common runtime arguments used for creation of DistributionSpec.
+ * The order of arguments in the args array is: [rank, num_banks, tensor_shape, shard_shape, bank_coords].
+ *
+ * @tparam CTA_OFFSET_  base index of compile-time arguments in the args array
+ * @tparam CRTA_OFFSET_ base index of common runtime arguments in the args array, if set to UNKNOWN, it will be set in
+ * constructor
+ */
+template <size_t CTA_OFFSET_, size_t CRTA_OFFSET_ = UNKNOWN>
+struct ArgsOffsets {
+    static constexpr size_t CTA_OFFSET = CTA_OFFSET_;
+    static constexpr size_t CRTA_OFFSET = CRTA_OFFSET_;
+
+    static constexpr auto args_config =
+        ArgsConfig(static_cast<ArgsConfig::Underlying>(get_compile_time_arg_val(CTA_OFFSET)));
+
+    using ArgsLoc = ArgsLocation<
+        args_config.test(ArgConfig::RankCRTA),
+        args_config.test(ArgConfig::NumBanksCRTA),
+        args_config.test(ArgConfig::TensorShapeCRTA),
+        args_config.test(ArgConfig::ShardShapeCRTA),
+        args_config.test(ArgConfig::BankCoordsCRTA)>;
+
+    // Impossible to have runtime rank without runtime tensor and shard shapes since then impossible to calculate CTA
+    // offsets in compile time
+    static_assert(
+        !ArgsLoc::RankCRTA or (ArgsLoc::RankCRTA and ArgsLoc::TensorShapeCRTA and ArgsLoc::ShardShapeCRTA),
+        "If rank is runtime, tensor_shape and shard_shape must also be runtime");
+    static_assert(
+        !ArgsLoc::NumBanksCRTA or (ArgsLoc::NumBanksCRTA and ArgsLoc::BankCoordsCRTA),
+        "If num_banks is runtime, bank_coords must also be runtime");
+
+    // Calculate offsets for compile-time arguments
+    static constexpr uint32_t ArgsConfigCTAOFfset = CTA_OFFSET;
+    static constexpr uint32_t RankCTAOffset = ArgsConfigCTAOFfset + 1;
+    static constexpr uint32_t NumBanksCTAOffset = RankCTAOffset + (ArgsLoc::RankCRTA ? 0 : 1);
+
+    static constexpr uint32_t RankCT =
+        ArgsLoc::RankCRTA ? 0 : get_compile_time_arg_val(ArgsLoc::RankCRTA ? CTA_OFFSET : RankCTAOffset);
```

Comment:
```suggestion
    static constexpr uint32_t RankCT = [] {
        if constexpr (rank_is_crta) {
            return 0;
        } else {
            return get_compile_time_arg_val(RankCTAOffset);
        }
    }();
```
This is better as 
- Remove redundant condition (rank_is_crta used twice).
- Remove risk of get_compile_time_arg_val misuse.

Reply:
Nice, I like it